src/prepared_statement.c

```diff
@@ -0,0 +1,388 @@
+	client->pool->stats.ps_bind_count++;
+
+	HASH_FIND_STR(client->client_prepared_statements, bp.name, client_ps);
+	if (!client_ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found", bp.name);
+		return false;
+	}
+	ps = client_ps->ps;
+
+	HASH_FIND_INT(server->server_prepared_statements, &ps->query_id, server_ps);
+	if (!server_ps) {
+		/* Statement is not prepared on this link, sent P packet first */
+		slog_debug(server, "handle_bind_command: prepared statement '%s' (query '%s') not available on server, preparing '%s' before bind",
+				   bp.name, ps->query_with_params, ps->stmt_name);
+
+		extract_parse_packet(ps, &pp);
+		buf = create_parse_packet(ps->stmt_name, &pp);
+
+		/* update stats */
+		client->pool->stats.ps_server_parse_count++;
+
+		/* Track Parse command sent to server */
+		opp = calloc(sizeof *opp, 1);
+		opp->ignore = true;
+		list_append(&server->server_outstanding_parse_packets, &opp->node);
+		if (!pktbuf_send_queued(buf, server)) {
+			return false;
+		}
+
+		/* Register statement on server link */
+		server_ps = create_server_prepared_statement(ps);
+		if (!register_prepared_statement(server, server_ps))
+			return false;
+	}
+	else if (HASH_COUNT(server->server_prepared_statements) != 1) /* do nothing if there is just one prepared statement */
+	{
+		/* Maintain LRU double-linked list: move element to the end of the list */
+		HASH_UNLINK(hh, server->server_prepared_statements, server_ps);
+		HASH_APPEND_LIST(hh, server->server_prepared_statements, server_ps);
+	}
+
+	slog_debug(client, "handle_bind_command: mapped statement '%s' (query '%s') to '%s'",
+			   pp.name, ps->query_with_params, ps->stmt_name);
+
+	diff = strlen(bp.name) - ps->stmt_name_len;
+	len = pkt->len - diff;
+	buf = pktbuf_dynamic(len);
+	pktbuf_put_char(buf, pkt->type);
+	pktbuf_put_uint32(buf, len - 1); /* length does not include type byte */
+	pktbuf_put_string(buf, bp.portal);
+	pktbuf_put_string(buf, ps->stmt_name);
+	pktbuf_put_bytes(buf, pkt->data.data + buf->write_pos + diff, len - buf->write_pos);
+	if (!pktbuf_send_queued(buf, server)) {
+		pktbuf_free(buf);
+		return false;
+	}
+	return true;
+}
+
+bool handle_describe_command(PgSocket *client, PktHdr *pkt)
+{
+	SBuf *sbuf = &client->sbuf;
+	PgSocket *server = client->link;
+	PgDescribePacket dp;
+	PgClientPreparedStatement *client_ps = NULL;
+	PgServerPreparedStatement *server_ps = NULL;
+	PgPreparedStatement* ps;
+	PktBuf *buf;
+	bool res;
+
+	Assert(server);
+
+	if (!unmarshall_describe_packet(client, pkt, &dp) || dp.type != 'S')
+		return false;
+
+	HASH_FIND_STR(client->client_prepared_statements, dp.name, client_ps);
+	if (!client_ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found", dp.name);
+		return false;
+	}
+	ps = client_ps->ps;
+
+	HASH_FIND_INT(server->server_prepared_statements, &ps->query_id, server_ps);
+	if (!server_ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found at server", dp.name);
```

Comment:
This should not disconnect. It should instead prepare the statement on the server, just like we are doing for Bind.

Reply:
Done

Reply:
My point is this should not be an error at all. Afaict it's completely valid to do following:
1. Begin transaction
2. Parse p1
4. Commit
5. Begin transaction
6. Describe p1
7. Commit

So it's completely possible that the prepared statement that the client wants to be described is not available on the server that is currently linked. Just like it is for Bind. So we should do the same that we do for bind, and prepare the statement on the linked server if it's not prepared there yet.

Reply:
Agree.
Replaced fatal with sending error to the client.

Reply:
I think you misunderstood what I meant. It's not an error at all. For Describe when the prepared statement is not yet prepared on the linked server (but is prepared on the client), we should simply prepare it. Just like we do for Bind: when we receive a Bind (for a prepared statement that's prepared by the client) we prepare the prepared statement on the linked server if it's not yet prepared there.

Reply:
I think that it is not possible to describe statement prepared in other transaction.
At least it was stated in the original code.
But I didn't find any prove of it in protocol description.
In any case - it seems to be better to prepare statement and return valid response  rather then throw error.
So I changed `handle_describe_command` to do the same as `handle_bind_command`